Convert app to Kotlin

maniac103 opened the pull request:
Since Google announced Android development will become Kotlin first [1] I guess it's a good time to convert the app to Kotlin ;-)

[1] https://android-developers.googleblog.com/2019/05/google-io-2019-empowering-developers-to-build-experiences-on-Android-Play.html

maniac103 commented:
TODO:

- [ ] adapt contributing guid
- [ ] look into how to replace AsyncTask
- [ ] adjust style guide?

mueller-ma commented:
I haven't used Kotlin much so far, but IMO this is the right direction.
How should we handle the merge of this PR? Merge a few pending PRs (#1358, #1357, #1338) first and then this one?

maniac103 commented:
> Merge a few pending PRs (#1358, #1357, #1338) first and then this one?

Fine with me ... as long as I don't have to rebase endlessly everything's fine :-)

I'll look into reviewing the mentioned 3 PRs and rebase when they're merged.

maniac103 commented:
- [ ] Cloud notification list crashes: `05-24 19:51:16.527 E/AndroidRuntime(18884): kotlin.UninitializedPropertyAccessException: lateinit property emptyView has not been initialized`

maniac103 commented on mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt:
```diff
@@ -0,0 +1,366 @@
+            Toasty.success(context, R.string.home_shortcut_success_pinning).show();
+        } else {
+            Toasty.error(context, R.string.home_shortcut_error_pinning).show();
+        }
```
Probably need to show the toast in main thread.

mueller-ma commented on mobile/src/main/java/org/openhab/habdroid/ui/CloudNotificationListFragment.kt:
```diff
@@ -61,6 +61,9 @@ class CloudNotificationListFragment : Fragment(), View.OnClickListener, SwipeRef
         val view = inflater.inflate(R.layout.fragment_notificationlist, container, false)
 
         recyclerView = view.findViewById(android.R.id.list)
+        emptyView = view.findViewById(android.R.id.empty)
```
Couldn't you use https://kotlinlang.org/docs/tutorials/android-plugin.html#view-binding here?

maniac103 replied:
I started doing this, but the imports due to this technique got totally out of hand in WidgetAdapter due to the large number of different layouts involved. I therefore didn't use it there and for consistency reasons didn't use it elsewhere either.

Additionally, there are at least some people at Google who think it's a bad idea to use it. Interesting read in that regard: https://proandroiddev.com/the-argument-over-kotlin-synthetics-735305dd4ed0

mueller-ma commented on mobile/src/foss/java/org/openhab/habdroid/ui/MapViewHelper.kt:
```diff
@@ -0,0 +1,219 @@
+    }
+}
+
+fun MapView.setMarker(pos: GeoPoint, item: Item, label: CharSequence?, canDrag: Boolean,
```
`MapView` can be null here. The current master crashes because of this.

maniac103 replied:
How? It's initialized in the constructor or created explicitly for the dialog.

mueller-ma replied:
Nevermind, the crash is related to an internal osm lib bug.

mueller-ma commented:
My AS claimed, that Kotlin isn't configured. I had to change the Kotlin release branch under "Tools" > "Kotlin". IMO this should be described in Readme.

mueller-ma commented:
For Stickler: https://stickler-ci.com/docs#kotlin

maniac103 commented:
> My AS claimed, that Kotlin isn't configured. I had to change the Kotlin release branch under "Tools" > "Kotlin".

This does not seem to be repo specific, but rather AS install specific, though? At least I did *not* have this problem with a fresh checkout of the kotlin branch in a fresh clone of my repo.

mueller-ma commented:
Yes, this is AS specific, but IMO the Readme should guide from a default AS installation to state where this project can be build.
Did you build any other Kotlin project before?

maniac103 commented:
> Did you build any other Kotlin project before?

No, but I vaguely recollect I may have gotten this message as well when I started with the Kotlin conversion. I don't remember exactly though; I think I shrugged it off because Kotlin wasn't configured properly in the project at that time either.

mueller-ma commented:
Crash when trying to open local url preference if no url is set:
````
org.openhab.habdroid E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.openhab.habdroid, PID: 8213
    java.lang.IllegalArgumentException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkParameterIsNotNull, parameter state
        at org.openhab.habdroid.ui.widget.UrlInputPreference.showDialog(Unknown Source:2)
        at android.preference.DialogPreference.onClick(DialogPreference.java:277)
        at android.preference.Preference.performClick(Preference.java:1135)
        at android.preference.PreferenceScreen.onItemClick(PreferenceScreen.java:249)
        at android.widget.AdapterView.performItemClick(AdapterView.java:318)
        at android.widget.AbsListView.performItemClick(AbsListView.java:1159)
        at android.widget.AbsListView$PerformClick.run(AbsListView.java:3136)
        at android.widget.AbsListView$3.run(AbsListView.java:4052)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
````

maniac103 commented:
> Crash when trying to open local url preference if no url is set:

Thanks, fixed in https://github.com/openhab/openhab-android/pull/1356/commits/f17238b9e6f5352b78011f9f218eba9a47fb19f4

maniac103 commented:
I have some follow-up changes pending which clean up the code even further, but I'm not sure whether they should be moved into separate PRs:
- https://github.com/maniac103/openhab.android/commit/15101f9238ff4096dd56a8f8f60173f5c5affe66 (maybe could inline into this one?)
- https://github.com/maniac103/openhab.android/commit/ad84c04286d2c69b7a5eb10c3a315d44c0627f7f (probably worth a separate PR)

mueller-ma commented:
>  maniac103@15101f9 (maybe could inline into this one?)
>  maniac103@ad84c04 (probably worth a separate PR)

Agreed.

mueller-ma commented:
On crashlytics I see the following crash in WidgetImageView.kt:
````
Fatal Exception: android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views.
       at android.view.ViewRootImpl.checkThread(ViewRootImpl.java:7760)
       at android.view.ViewRootImpl.requestLayout(ViewRootImpl.java:1225)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at androidx.drawerlayout.widget.DrawerLayout.requestLayout(DrawerLayout.java:1303)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at androidx.recyclerview.widget.RecyclerView.requestLayout(RecyclerView.java:4202)
       at android.view.View.requestLayout(View.java:23093)
       at android.view.View.requestLayout(View.java:23093)
       at android.widget.ImageView.setScaleType(ImageView.java:820)
       at org.openhab.habdroid.ui.widget.WidgetImageView.removeProgressDrawable(WidgetImageView.kt:205)
       at org.openhab.habdroid.ui.widget.WidgetImageView.access$removeProgressDrawable(WidgetImageView.kt:29)
       at org.openhab.habdroid.ui.widget.WidgetImageView$HttpImageRequest$execute$1.invokeSuspend(WidgetImageView.kt:234)
       at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
       at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:285)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:594)
       at kotlinx.coroutines.scheduling.CoroutineScheduler.access$runSafely(CoroutineScheduler.kt:60)
       at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:742)
````

maniac103 commented:
That one is already fixed, I noticed it myself :-)

mueller-ma commented:
Can you "fix" the issue found by LGTM? https://lgtm.com/projects/g/openhab/openhab-android/alerts/?mode=list
`item` can be null here, but only if `success` is false, so it's actually never called while being null.

mueller-ma commented on mobile/src/main/java/org/openhab/habdroid/ui/MainActivity.kt:
```diff
@@ -0,0 +1,1020 @@
+        drawerToggle.isDrawerIndicatorEnabled = !controller.canGoBack()
+    }
+
+    override fun onBackPressed() {
```
IMO override methods aka "Android framework methods" should be at the beginning of a class and all "custom/own" functions after them.

mueller-ma commented on mobile/src/main/java/org/openhab/habdroid/ui/WidgetListFragment.kt:
```diff
@@ -113,11 +109,11 @@ class WidgetListFragment : Fragment(), WidgetAdapter.ItemClickListener {
                     } else {
-                        val itemToHandle = if (which < commands.size) widget.item else null
-                        val linkToHandle = if (which == commands.size) widget.linkedPage?.link else null
+                        val itemToHandle = if (which < commands.size) item.item else null
```
IMO `item` should be renamed to `widget`. I know that two meanings of item conflict here (openHAB Item vs. thing that was clicked), but `item.item` seems too strange to me.

maniac103 replied:
I agree, but there's only suboptimal solutions here:
- suboptimal naming (what we have now)
- name the parameter `widget`, which yields a warning as the parameter is named `item` in the base class
- name the parameter `item` and create a variable `widget` from it, which yields a warning because the variable is redundant

:-/

mueller-ma replied:
I like option 2 and suppressing the warning most.

maniac103 replied:
I totally forgot that listener is our code as well ... that makes fixing this pretty easy actually: https://github.com/openhab/openhab-android/pull/1356/commits/b30e0912b1a2f48a905debffb0f4391f4ae0b088

maniac103 commented:
> Can you "fix" the issue found by LGTM? https://lgtm.com/projects/g/openhab/openhab-android/alerts/?mode=list
> item can be null here, but only if success is false, so it's actually never called while being null.

I'll look into improving this.

> By the way: I've installed LGTM as GitHub App now. See #1386. Can you close and reopen this PR, so LGTM can scan for new issues?

Sure.

mueller-ma commented:
> Can you close and reopen this PR, so LGTM can scan for new issues?

Kotlin is currently unsupported: https://discuss.lgtm.com/t/kotlin-native-support/1636

maniac103 commented:
> Thanks, fixed in f17238b

Same issue still exists at https://github.com/maniac103/openhab.android/blob/kotlin-experimental/mobile/src/main/java/org/openhab/habdroid/ui/widget/ItemUpdatingPreference.kt#L90

maniac103 commented on mobile/src/main/java/org/openhab/habdroid/util/HttpClient.kt:
```diff
@@ -0,0 +1,211 @@
+                }
+            }
+            val result = HttpBitmapResult(request, bitmap)
+            close()
```
The close() calls need to happen on IO thread.

mueller-ma commented on mobile/src/main/java/org/openhab/habdroid/ui/ItemPickerActivity.kt:
```diff
@@ -91,7 +91,7 @@ class ItemPickerActivity : AbstractBaseActivity(), SwipeRefreshLayout.OnRefreshL
         if (!getPrefs().isTaskerPluginEnabled()) {
             isDisabled = true
-            updateViewVisibility(false, false, true)
+            updateViewVisibility(loading = false, loadError = false, isDisabled = true)
```
AS is printing the parameter name next to the arguments, if these are primitive types. So IMO this change isn't really needed.

maniac103 replied:
There's a style inspection that emits a warning if there's multiple unnamed boolean parameters for one method, presumably to avoid them being confused.

mueller-ma commented:
Crash after app start:
````
java.lang.RuntimeException: Unable to create application org.openhab.habdroid.core.OpenHabApplication: kotlin.UninitializedPropertyAccessException: lateinit property instance has not been initialized
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4883)
        at android.app.ActivityThread.access$1500(ActivityThread.java:160)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1382)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5597)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:984)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
     Caused by: kotlin.UninitializedPropertyAccessException: lateinit property instance has not been initialized
        at org.openhab.habdroid.util.CacheManager$Companion.getInstance(CacheManager.kt:62)
        at org.openhab.habdroid.core.connection.ConnectionFactory.<init>(ConnectionFactory.kt:72)
        at org.openhab.habdroid.core.connection.ConnectionFactory$Companion.initialize(ConnectionFactory.kt:414)
        at org.openhab.habdroid.core.OpenHabApplication.onCreate(OpenHabApplication.kt:12)
        at android.app.Instrumentation.callApplicationOnCreate(Instrumentation.java:1015)
        at android.app.ActivityThread.handleBindApplication(ActivityThread.java:4880)
        at android.app.ActivityThread.access$1500(ActivityThread.java:160) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1382) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5597) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:984) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779) 
````

mueller-ma commented:
Crash when opening settings:
````
06-11 17:20:22.188 org.openhab.habdroid.beta E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.openhab.habdroid.beta, PID: 8108
    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.openhab.habdroid.beta/org.openhab.habdroid.ui.PreferencesActivity}: java.lang.IllegalArgumentException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkParameterIsNotNull, parameter defaultValue
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2473)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2541)
        at android.app.ActivityThread.access$800(ActivityThread.java:160)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5597)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:984)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
     Caused by: java.lang.IllegalArgumentException: Parameter specified as non-null is null: method kotlin.jvm.internal.Intrinsics.checkParameterIsNotNull, parameter defaultValue
        at org.openhab.habdroid.ui.widget.ItemUpdatingPreference.onSetInitialValue(ItemUpdatingPreference.kt)
        at android.preference.Preference.dispatchSetInitialValue(Preference.java:1377)
        at android.preference.Preference.onAttachedToHierarchy(Preference.java:1172)
        at android.preference.PreferenceGroup.addPreference(PreferenceGroup.java:167)
        at android.preference.PreferenceGroup.addItemFromInflater(PreferenceGroup.java:108)
        at android.preference.PreferenceGroup.addItemFromInflater(PreferenceGroup.java:45)
        at android.preference.GenericInflater.rInflate(GenericInflater.java:488)
        at android.preference.GenericInflater.rInflate(GenericInflater.java:493)
        at android.preference.GenericInflater.inflate(GenericInflater.java:326)
        at android.preference.GenericInflater.inflate(GenericInflater.java:263)
        at android.preference.PreferenceManager.inflateFromResource(PreferenceManager.java:272)
        at android.preference.PreferenceFragment.addPreferencesFromResource(PreferenceFragment.java:299)
        at org.openhab.habdroid.ui.PreferencesActivity$MainSettingsFragment.updateAndInitPreferences(PreferencesActivity.kt:184)
        at org.openhab.habdroid.ui.PreferencesActivity$AbstractSettingsFragment.onCreate(PreferencesActivity.kt:112)
        at android.app.Fragment.performCreate(Fragment.java:2031)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:863)
        at android.app.FragmentManagerImpl.moveToState(FragmentManager.java:1067)
        at android.app.BackStackRecord.run(BackStackRecord.java:834)
        at android.app.FragmentManagerImpl.execPendingActions(FragmentManager.java:1452)
        at android.app.Activity.performStart(Activity.java:6041)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2436)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2541) 
        at android.app.ActivityThread.access$800(ActivityThread.java:160) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1321) 
        at android.os.Handler.dispatchMessage(Handler.java:102) 
        at android.os.Looper.loop(Looper.java:135) 
        at android.app.ActivityThread.main(ActivityThread.java:5597) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at java.lang.reflect.Method.invoke(Method.java:372) 
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:984) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779) 
````

maniac103 commented:
Fixed both. Sorry, should have tried at least the former, the latter wasn't a regression from earlier code.

maniac103 commented:
There seems to be an issue with image loading (WidgetImageView), otherwise this starts to look in good shape.

mueller-ma commented:
and:

> review whether some fields in data model could be made non-nullable

When it's RTM from your side, I'll run some monkey tests to (maybe) find some crashes.

maniac103 commented on mobile/src/main/java/org/openhab/habdroid/model/LinkedPage.kt:
```diff
@@ -0,0 +1,67 @@
+
+@Parcelize
+data class LinkedPage(val id: String?, val title: String?, val icon: String?,
+                      val iconPath: String, val link: String) : Parcelable {
```
Are `id` and `title` really nullable (IOW, optional)? It's been like that since [like forever](https://github.com/openhab/openhab-android/blob/7aa2593b9e6f41d7c479b4c29879b7ede0fcf801/mobile/src/main/java/org/openhab/habdroid/model/OpenHABLinkedPage.java#L52-L59), but it doesn't look correct to me at least for for id on OH2.

mueller-ma replied:
None of these make sense to be nullable. I suggest we make it non-nullable and look for crash reports.

maniac103 commented on mobile/src/main/java/org/openhab/habdroid/model/Item.kt:
```diff
@@ -0,0 +1,149 @@
+
+@Parcelize
+data class Item internal constructor(val name: String, val label: String?, val category: String?,
+                                     val type: Type, val groupType: Type?, val link: String?,
```
Is `link` actually optional?

maniac103 replied:
Looks like it can be optional, in particular for items sent via SSE.

maniac103 commented on mobile/src/main/java/org/openhab/habdroid/model/Sitemap.kt:
```diff
@@ -0,0 +1,93 @@
+import org.w3c.dom.Node
+
+@Parcelize
+data class Sitemap internal constructor(val name: String, val label: String, val link: String?,
```
Is `link` actually optional?

maniac103 commented:
> For Stickler: https://stickler-ci.com/docs#kotlin

ktlint seems to be a bit tricky, in that it is [not configurable on purpose](https://github.com/pinterest/ktlint/blob/master/README.md) and doesn't match default AS configuration, in particular about parameter indention (parameter list indention is expected to be 4 spaces, while AS aligns). ktlint follows official style guide, but I guess we need to look into including AS config in our repo for that. ktlint has a single line command to apply the config changes to the project files, but I don't think we can expect our users to all install ktlint.

mueller-ma commented:
Including parts of the `.idea` directory is definitely a good idea, also for stuff like import order. This can be part of a different PR, though.
You can keep Stickler as is in this PR.

maniac103 commented:
`ktlint` is happy now, but `ktlint --android` is not, mainly because I increased max. line length to 120. Rationale for that was that with the tendentially longer lines in Kotlin the default line length of 100 would lead to *a lot* of extra line wrapping (currently counting 220 lines > 100 chars) ... thoughts about this? Options are accepting the wrapping or shipping a `.editorconfig` file. I think I prefer the latter ;-)

maniac103 commented:
I've replaced checkstyle with ktlint for Stickler, and the only problem that I can see is the ktlint version used by Stickler is too old and complains about the wildcard imports. The latest ktlint version [doesn't do that (anymore)](https://github.com/pinterest/ktlint/pull/435).

mueller-ma commented on .stickler.yml:
```diff
@@ -1,5 +1,6 @@
 linters:
-    checkstyle:
```
I'd like to keep checkstyle for the MTM lib.

mueller-ma commented:
> I think I prefer the latter ;-)

Indeed.

Stickler uses 0.31.0, while the latest version is https://github.com/pinterest/ktlint/releases/tag/0.33.0. I'll write the support a mail.

mueller-ma commented:
If I use a home screen shortcut to open the app, the sitemap page isn't opened correctly. It shows "This page contains no visible widgets".

maniac103 commented:
> If I use a home screen shortcut to open the app, the sitemap page isn't opened correctly. It shows "This page contains no visible widgets".

Already found this myself ;-) Fixed.

mueller-ma commented:
Two things:
1. Crash when setting item update prefix, leaving and re-entering prefs:
````java
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.openhab.habdroid.beta, PID: 5622
    kotlin.KotlinNullPointerException
        at org.openhab.habdroid.ui.WidgetListFragment.onViewCreated(WidgetListFragment.kt:91)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1471)
        at androidx.fragment.app.FragmentManagerImpl.moveFragmentToExpectedState(FragmentManager.java:1784)
        at androidx.fragment.app.FragmentManagerImpl.moveToState(FragmentManager.java:1852)
        at androidx.fragment.app.FragmentManagerImpl.dispatchStateChange(FragmentManager.java:3269)
        at androidx.fragment.app.FragmentManagerImpl.dispatchActivityCreated(FragmentManager.java:3229)
        at androidx.fragment.app.FragmentController.dispatchActivityCreated(FragmentController.java:201)
        at androidx.fragment.app.FragmentActivity.onStart(FragmentActivity.java:620)
        at androidx.appcompat.app.AppCompatActivity.onStart(AppCompatActivity.java:178)
        at org.openhab.habdroid.ui.MainActivity.onStart(MainActivity.kt:233)
        at android.app.Instrumentation.callActivityOnStart(Instrumentation.java:1391)
        at android.app.Activity.performStart(Activity.java:7157)
        at android.app.ActivityThread.handleStartActivity(ActivityThread.java:2937)
        at android.app.servertransaction.TransactionExecutor.performLifecycleSequence(TransactionExecutor.java:180)
        at android.app.servertransaction.TransactionExecutor.cycleToPath(TransactionExecutor.java:165)
        at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:142)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:70)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1808)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
````

2. HABPanel doesn't load. I only see the loading indicator.

mueller-ma commented:
Crash when clicking on OSM MapView:
````
06-19 18:07:24.751 org.openhab.habdroid.beta E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.openhab.habdroid.beta, PID: 8481
    java.lang.NullPointerException: Attempt to invoke virtual method 'boolean android.os.Handler.post(java.lang.Runnable)' on a null object reference
        at org.openhab.habdroid.ui.MapViewHelper$OsmViewHolder.openPopup(MapViewHelper.kt:143)
        at org.openhab.habdroid.ui.MapViewHelper$OsmViewHolder.access$openPopup(MapViewHelper.kt:45)
        at org.openhab.habdroid.ui.MapViewHelper$OsmViewHolder$$special$$inlined$with$lambda$1.singleTapConfirmedHelper(MapViewHelper.kt:67)
        at org.osmdroid.views.overlay.MapEventsOverlay.onSingleTapConfirmed(MapEventsOverlay.java:39)
        at org.osmdroid.views.overlay.DefaultOverlayManager.onSingleTapConfirmed(DefaultOverlayManager.java:305)
        at org.osmdroid.views.MapView$MapViewDoubleClickListener.onSingleTapConfirmed(MapView.java:1594)
        at android.view.GestureDetector$GestureHandler.handleMessage(GestureDetector.java:273)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:135)
        at android.app.ActivityThread.main(ActivityThread.java:5597)
        at java.lang.reflect.Method.invoke(Native Method)
        at java.lang.reflect.Method.invoke(Method.java:372)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:984)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:779)
````

mueller-ma commented:
A few lint issue fixes in foss MapView: https://github.com/mueller-ma/openhab.android/commit/adb1736a7ca1e70ccec60a458e53d297d518e1da

maniac103 commented:
> A few lint issue fixes in foss MapView: mueller-ma@adb1736

Picked.

> Crash when clicking on OSM MapView:

Fixed.

> HABPanel doesn't load. I only see the loading indicator.

Fixed.

> Crash when setting item update prefix, leaving and re-entering prefs:

This one is interesting. It doesn't seem to be related to the prefs, but to a situation where a widget list fragment is present without any active connection. Can you reproduce this at will, and/or was there any connection change while you were testing this? I don't think this issue is related to the Kotlin conversion, as the basic assumption in code 'if widget list is shown, connection must be present, as otherwise activity would have removed the fragment' existed before as well. It might be that timing has changed now, as Kotlin will check for non-null immediately, while Java only checks when dereferencing the object.

mueller-ma commented:
> Picked.

Forgot to push?

maniac103 commented:
> Forgot to push?

Indeed, it's there now.

mueller-ma commented:
RTM from your side?

maniac103 commented:
Yeah :+1:

mueller-ma commented:
By the way Stickler updated ktlint.

mueller-ma commented:
````
> Task :mobile:testFossBetaReleaseUnitTest
org.openhab.habdroid.core.connection.ConnectionFactoryTest > testGetAnyConnectionWifiLocalRemote FAILED
    kotlin.KotlinNullPointerException at ConnectionFactoryTest.kt:205
73 tests completed, 1 failed
> Task :mobile:testFossBetaReleaseUnitTest FAILED
FAILURE: Build failed with an exception.
````

maniac103 commented:
Wat ... line 205 is `runBlocking {`, how can that throw a NPE? I'll check it again, but I ran the test suite various times and didn't see anything like that.

mueller-ma commented:
````
> Task :mobile:testFullStableReleaseUnitTest
org.openhab.habdroid.core.connection.ConnectionFactoryTest > testGetAnyConnectionUnsupportedNetwork FAILED
    java.lang.Exception
        Caused by: kotlin.KotlinNullPointerException at ConnectionFactoryTest.kt:165
73 tests completed, 1 failed
````
from https://travis-ci.org/openhab/openhab-android/builds/548904166#L2152

maniac103 commented:
Line 165 also is runBlocking ... I wonder whether something's off in their JVM setup or something ... I guess we need to ask their support about that.

mueller-ma commented:
Can this be related to a bad internet connection?

maniac103 commented:
I think the runBlocking function probably catches and rethrows all exceptions thrown in its scope. I'll check whether that's the case and how we could get to the actual cause.

mueller-ma commented:
````
org.openhab.habdroid.beta E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.openhab.habdroid.beta, PID: 12328
    kotlin.UninitializedPropertyAccessException: lateinit property helpIcon has not been initialized
        at org.openhab.habdroid.ui.widget.ItemUpdatingPreference.onCheckedChanged(ItemUpdatingPreference.kt:117)
        at android.widget.CompoundButton.setChecked(CompoundButton.java:154)
        at androidx.appcompat.widget.SwitchCompat.setChecked(SwitchCompat.java:1060)
        at org.openhab.habdroid.ui.widget.ItemUpdatingPreference.onCreateDialogView(ItemUpdatingPreference.kt:80)
        at android.preference.DialogPreference.showDialog(DialogPreference.java:295)
        at org.openhab.habdroid.ui.widget.ItemUpdatingPreference.showDialog(ItemUpdatingPreference.kt:93)
````

maniac103 commented:
Oops, a crash that's only reproducible if the pref is already set :-/ Lines 84-85 need to be moved up; I'll create a PR.

maniac103 commented:
> Task :mobile:testFullStableReleaseUnitTest
org.openhab.habdroid.core.connection.ConnectionFactoryTest > testGetAnyConnectionUnsupportedNetwork FAILED
    java.lang.Exception
        Caused by: kotlin.KotlinNullPointerException at ConnectionFactoryTest.kt:165
73 tests completed, 1 failed

Actual NPE is [here](https://github.com/openhab/openhab-android/pull/1356/files#diff-d15176dc32c614d962a77bf7be5f19f6R474) ... I'll look into whether there are race conditions in that init code.
